hazelcast/src/main/java/com/hazelcast/map/impl/event/MapEventPublisherImpl.java

```diff
@@ -140,75 +158,114 @@ public void publishEvent(Address caller, String mapName, EntryEventType eventTyp
+
+        Map<Integer, EntryEventData> eventDataIncludingValues = null;
+        Map<Integer, EntryEventData> eventDataExcludingValues = null;
+        int orderKey = pickOrderKey(dataKey);
 
         for (EventRegistration registration : registrations) {
             EventFilter filter = registration.getFilter();
-
-            if (!doFilter(filter, dataKey, oldValue, value, eventType, mapName)) {
+            // a filtering strategy determines whether the event must be published on the specific
+            // event registration and may alter the type of event to be published
+            int eventTypeForPublishing = filteringStrategy.doFilter(filter, dataKey, oldValue, value, eventType, mapName);
+            if (eventTypeForPublishing == FILTER_DOES_NOT_MATCH) {
                 continue;
             }
 
+            EntryEventData eventDataToBePublished;
             if (isIncludeValue(filter)) {
-                includeValueRegistrations = getOrCreateList(includeValueRegistrations);
-                includeValueRegistrations.add(registration);
+                if (eventDataIncludingValues == null) {
+                    eventDataIncludingValues = new Int2ObjectHashMap<EntryEventData>(EVENT_DATA_MAP_CAPACITY);
+                }
+                eventDataToBePublished = getOrCreateEventData(eventDataIncludingValues, mapName, caller, dataKey,
+                        value, oldValue, mergingValue, eventTypeForPublishing);
             } else {
-                nullValueRegistrations = getOrCreateList(nullValueRegistrations);
-                nullValueRegistrations.add(registration);
+                if (eventDataExcludingValues == null) {
+                    eventDataExcludingValues = new Int2ObjectHashMap<EntryEventData>(EVENT_DATA_MAP_CAPACITY);
+                }
+                eventDataToBePublished = getOrCreateEventData(eventDataExcludingValues, mapName, caller, dataKey,
+                        null, null, null, eventTypeForPublishing);
             }
+            eventService.publishEvent(SERVICE_NAME, registration, eventDataToBePublished, orderKey);
         }
 
-        if (!isEmpty(includeValueRegistrations)) {
-            Data dataOldValue = mapServiceContext.toData(oldValue);
-            Data dataValue = mapServiceContext.toData(value);
-            Data dataMergingValue = mapServiceContext.toData(mergingValue);
-            EntryEventData eventData = createEntryEventData(mapName, caller, dataKey,
-                    dataValue, dataOldValue, dataMergingValue, eventType.getType());
-            int orderKey = pickOrderKey(dataKey);
-            publishEventInternal(includeValueRegistrations, eventData, orderKey);
+        // if events were generated, execute the post-publish hook on each one
+        if (eventDataIncludingValues != null) {
+            for (EntryEventData entryEventData : eventDataIncludingValues.values()) {
+                postPublishEvent(entryEventData);
+            }
```

Comment:
To prevent unnecessary iterations, signature can be changed to receive a collection `postPublishEvent(Collection<EntryEventData> eventData)`